### PR TITLE
Listen on ETW and Wirte to STDOUT

### PR DESCRIPTION
**Pull request by mcy94w**
ETW Listener will Listen on the IIS Logging ETW Channel { 0x7E8AD27F, 0xB271, 0x4EA2,{ 0xA7, 0x83, 0xA4, 0x7B, 0xDE, 0x29, 0x14, 0x3B } } and write every etw event to STDOUT.

### CODE REVIEW

**Comment by mcy94w** [comment]
https://github.com/Microsoft/IIS.ServiceMonitor/issues/1

**Comment by shirhatti** [comment]
Can you add a comment with what the sample output looks like?

**Comment by mcy94w** [comment]
Service 'w3svc' has been stopped
                                                                                                                               NAMES
 Service 'w3svc' started                                                                                                /tcp   clever_meninsky
BYE
date 2017-10-05 time 18:15:00 s-sitename W3SVC1 s-computername ed4d698ac001 s-ip 172.23.82.122 cs-method GET cs-uri-stem /BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB cs-uri-query - port 80 cs-user-name - c-ip 10.127.64.214 cs-version HTTP/1.1 cs(User-Agent) - cs(Cookie) - cs(Referer) - cs-host 10.209.34.86 sc-status 404 sc-substatus 0 sc-win32-status 2 sc-bytes 1406 cs
-bytess 113 time-taken 6949 x-forwarded-for -                                                                                   NAMES
                                                                                                                        0/tcp   clever_meninsky
date 2017-10-05 time 18:15:15 s-sitename W3SVC1 s-computername ed4d698ac001 s-ip 172.23.82.122 cs-method GET cs-uri-stem /BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB cs-uri-query - port 80 cs-user-name - c-ip 10.127.64.214 cs-version HTTP/1.1 cs(User-Agent) - cs(Cookie) - cs(Referer) - cs-host 10.209.34.86 sc-status 404 sc-substatus 0 sc-win32-status 2 sc-bytes 1406 cs
-bytess 113 time-taken 98 x-forwarded-for -

date 2017-10-05 time 18:15:15 s-sitename W3SVC1 s-computername ed4d698ac001 s-ip 172.23.82.122 cs-method GET cs-uri-stem /BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB cs-uri-query - port 80 cs-user-name - c-ip 10.127.64.214 cs-version HTTP/1.1 cs(User-Agent) - cs(Cookie) - cs(Referer) - cs-host 10.209.34.86 sc-status 404 sc-substatus 0 sc-win32-status 2 sc-bytes 1406 cs
-bytess 113 time-taken 93 x-forwarded-for -

File: src/ServiceMonitor/EtwListner.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	CHAR *      userAgent         = ((CHAR *)port + USHORT_BYTES);
+	CHAR *      cookie            = ((CHAR *)userAgent + strlen(userAgent) + NULL_BYTES);
+	CHAR *      referer           = ((CHAR *)cookie + strlen(cookie) + NULL_BYTES);
+	WCHAR *     protocolVer       = (WCHAR *)(referer + strlen(referer) + CHAR_NULL_BYTES);
+	CHAR *      host              = (CHAR *)protocolVer + WCHAR_BYTES * wcslen(protocolVer) + WCHAR_NULL_BYTES;
+	USHORT *    protocolSubStatus = (USHORT *)(host + strlen(host) + CHAR_NULL_BYTES);
+	WCHAR *     customFields      = (WCHAR *)((CHAR *)protocolSubStatus + USHORT_BYTES);
+
+	/*
+	if ((customFields + wcslen(customFields)) <= userDataEnd)
```
Comment (jimmyca15) [inline src/ServiceMonitor/EtwListner.cpp]: Is this supposed to be here?

File: src/ServiceMonitor/EtwListner.cpp [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+EtwListner::~EtwListner()
+{
+}
+
+EtwListner::EtwListner()
+{
+	StartListen(IIS_LOG_TRACE, &ProviderGuid);
+}
+
+void EtwListner::StartListen(LPWSTR pStrSessionName, LPCGUID pTraceGUID)
```
[the remark this answers is not in the document]
Comment (mcy94w) [reply src/ServiceMonitor/EtwListner.cpp]: This feature is for debug container during development. To stop listen, will have to start a new container and not get log.

File: src/ServiceMonitor/EtwListner.cpp [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+#define CHAR_BYTES sizeof(CHAR)
+#define USHORT_BYTES sizeof(USHORT)
+#define ULONG_BYTESE sizeof(ULONG)
+#define ULONGLONG_BYTES sizeof(ULONGLONG)
+#define WCHAR_NULL_BYTES sizeof(WCHAR)
+#define CHAR_NULL_BYTES sizeof(CHAR)
+#define NULL_BYTES 1
+
+//0x7E8AD27F - 0xB271 - 0x4EA2 - 0xA783 - A47BDE29143B
+static const GUID ProviderGuid =
```
[the remark this answers is not in the document]
Comment (mcy94w) [reply src/ServiceMonitor/EtwListner.cpp]: This is the IIS GUID

File: src/ServiceMonitor/EtwListner.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#include <windows.h>
+
+
+class EtwListner
+{
+
+public:
+	EtwListner();
+	~EtwListner();
+	void StartListen(LPWSTR pStrSessionName, LPCGUID pTraceGUID);
```
[the remark this answers is not in the document]
Comment (mcy94w) [reply src/ServiceMonitor/EtwListner.h]: No need to have stop listen. We just wait for container to shut down.

File: src/ServiceMonitor/Main.cpp [hunk 5]
```diff
@@ -80,7 +87,14 @@ int __cdecl _tmain(int argc, _TCHAR* argv[])
...
     {
         hr = sm.StartServiceByName(argv[1]);
-		EtwListner sl = EtwListner();
+
+		//
+		// listen on IIS event and print the result to STDOUT
+		//
+		if (fListETW)
+		{
+			ewlList.StartListenIISEvent();
```
Comment (jimmyca15) [inline src/ServiceMonitor/Main.cpp]: Should start listen have a return value to check for success?

File: src/ServiceMonitor/EtwListner.cpp [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        0,
+        &dwThreadId);
+
+    if (pThreadHandle == NULL)
+    {
+        uStatus = GetLastError();
+        _tprintf(L"START THREAD FAIL %i\n", uStatus);
+        goto finish;
+    }
+    std::wcout << "BYE" << std::endl;
```
Comment (jimmyca15) [inline src/ServiceMonitor/EtwListner.cpp]: This trace should be more informative. _BYE_ can easily get lost within the console. If this trace is needed how about _Shutting down ETW Listener..._

File: src/ServiceMonitor/Main.cpp [hunk 7]
```diff
@@ -57,6 +59,11 @@ int __cdecl _tmain(int argc, _TCHAR* argv[])
         }
     }
 
+    if (argc >= 2 && (_wcsicmp(argv[1], L"-listETW") == 0))
```
Comment (jimmyca15) [inline src/ServiceMonitor/Main.cpp]: Is this short for listenETW?

File: src/ServiceMonitor/EtwListner.cpp (hunk 6 above)
Comment (mcy94w) [reply src/ServiceMonitor/EtwListner.cpp]: This is debug info. Will delete this.

**Comment by bariscaglar** [comment]
@mcy94w @shirhatti has this been documented?

**Comment by shirhatti** [comment]
No, this hasn't been documented yet.

**Comment by shirhatti** [comment]
Let's add an environment variable to toggle this as well `ENABLE_IIS_CONSOLE_LOGGING`

File: src/ServiceMonitor/IISConfigUtil.cpp [hunk 8]
```diff
@@ -54,25 +56,29 @@ IISConfigUtil::~IISConfigUtil()
     }
 }
 
-BOOL IISConfigUtil::FilterEnv(unordered_map<wstring, LPTSTR> filter, LPTSTR strEnvName, LPTSTR strEnvValue)
+BOOL IISConfigUtil::NameValuePairExist(unordered_map<wstring, wstring> filter, LPTSTR pStrName, LPTSTR pStrValue)
```
Comment (jimmyca15) [inline src/ServiceMonitor/IISConfigUtil.cpp]: Should be NameValuePairExist**s**

File: src/ServiceMonitor/IISConfigUtil.cpp [hunk 9]
```diff
@@ -5,8 +5,10 @@
 #include <iostream>
 using namespace std;
 
-#define KV(a,b) pair<wstring, LPTSTR>(a,b)
-#define KV_WSTR(a,b) pair<wstring, wstring>(a,b)
+#define KV(a,b) pair<wstring, wstring>(a,b)
+#define ENABLE_IIS_CONSOLE_LOGGING_NAME  L"ENABLE_IIS_CONSOLE_LOGGING"
+#define ENABLE_IIS_CONSOLE_LOGGING_VALUE L"1"
```
[the remark this answers is not in the document]
Comment (mcy94w) [reply src/ServiceMonitor/IISConfigUtil.cpp]: I don't think `ENABLED` nor `ENV_ENABLED ` contains enough information for this specific string.

**Pull request merged**